File: src/main/python/g0001_0100/s0015_3sum/Solution.py
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Sorting #Two_Pointers
# #Data_Structure_II_Day_1_Array #Algorithm_II_Day_3_Two_Pointers #Udemy_Two_Pointers
# #Top_Interview_150_Two_Pointers #Big_O_Time_O(n*log(n))_Space_O(n^2)
# #2024_06_04_Time_683_ms_(63.27%)_Space_20.8_MB_(49.61%)

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        length = len(nums)

        for i in range(length - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left = i + 1
            right = length - 1

            while left < right:
                total = nums[i] + nums[left] + nums[right]

                if total < 0:
                    left += 1
                elif total > 0:
                    right -= 1
                else:
                    result.append([nums[i], nums[left], nums[right]])

                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1

        return result

File: src/main/swift/g0001_0100/s0015_3sum/Solution.swift
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Sorting #Two_Pointers
// #Data_Structure_II_Day_1_Array #Algorithm_II_Day_3_Two_Pointers #Udemy_Two_Pointers
// #Top_Interview_150_Two_Pointers #Big_O_Time_O(n*log(n))_Space_O(n^2)
// #2024_06_18_Time_129_ms_(99.77%)_Space_20.6_MB_(57.21%)

class Solution {    
    func threeSum(_ nums: [Int]) -> [[Int]] {
        if nums.count < 3 { return [] }
        
        var nums = nums.sorted()
        guard var firstPositiveIndex = nums.firstIndex(where: { $0 >= 0 }) else { return [] }
        
        var result: [[Int]] = []
        var left = 0
        var right = nums.count - 1
        var amount = 0

        for i in (0...firstPositiveIndex) {
            if i > 0 && nums[i] == nums[i-1] { continue }
            
            left = i + 1
            right = nums.count - 1
            
            while left < right {
                
                var amount = nums[i] + nums[left] + nums[right]
                if amount == 0 {
                    if result.last != [nums[i], nums[left], nums[right]] {
                        result.append([nums[i], nums[left], nums[right]])
                    }
                    left += 1
                    right -= 1
                } else if amount > 0 {
                    right -= 1
                } else {
                    left += 1
                }
            }
        }

        return result
    }
}
